### PR TITLE
[Pipelining] Add schedule unshard/reshard pass

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131762
* #130860
* #130488
* #132369
* #130378
* __->__ #129810

Adds fsdp unshard/reshard ops to a compute-only schedule.

Operates on one pp-rank's schedule at a time, since there is no
cross-pp-rank coordination needed for FSDP.  (Unshard/Reshard is across
DP ranks within a PP group).

Uses a heuristic based on examining the next N stages to run compute
operations on this rank, evicting (resharding) and fetching (unsharding)
ahead of time to give unshard operations a chance to overlap with
compute and PP comms.
- this heuristic has not been validated and may not be optimal

Makes the assumption that it's fine to add the UNSHARD/RESHARD actions
to the schedule regardless of if FSDP will actually be used.
- this way, users do not have to tell us at PP schedule creation time if
  they plan to use FSDP or DDP
- it is trivial to implement UNSHARD/RESHARD as no-ops inside the
  runtime, if FSDP is not detected on the stage module

TODO
- also add FSDP's reduce-scatter? or is it sufficient to leave this
  handled by PipelineStage at 'last backward' time
- validate 'next N stages' heuristic and expose an API if needed
- add an e2e test

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/distributed/pipelining/schedules.py [hunk 1]
```diff
@@ -603,6 +603,135 @@ def _step_microbatches(
...
+        # instances of the same layer? hmm. that's not quite right either.
+        for a in next_actions:
+            if a is not None and a.stage_index not in seen:
+                seen.add(a.stage_index)
+                ret.append(a.stage_index)
+        return ret
+
+    def _per_rank_unshard(rank):
+        rank_actions = compute_actions[rank]
+        active_layers: Set[int] = set()
```
Comment (kwen2501) [inline torch/distributed/pipelining/schedules.py:685]: What's the definition of `layer` here? As like a real "layer" in a model or something else? Reads like a concrete thing.

File: torch/distributed/pipelining/schedules.py [hunk 2]
```diff
@@ -743,6 +753,135 @@ def _step_microbatches(
...
+                "Rank %d Step %d active: %s fetch %s, evict_candidates %s, num_evict = %d",
+                # TODO why is rank a str?
+                int(rank),
+                i,
+                active_layers,
+                fetch,
+                evict_candidates,
+                num_evict,
+            )
+            for layer in list(evict_candidates)[:num_evict]:
```
Comment (Skylion007) [inline torch/distributed/pipelining/schedules.py:874]: nit: you can probably just use itertools islice or something here.

File: torch/distributed/pipelining/schedules.py [hunk 3]
```diff
@@ -743,6 +753,135 @@ def _step_microbatches(
...
+            # fetch needs to be ordered correctly though, so don't use a set for it
+            fetch = [l for l in next_n[:max_active_layers] if l not in active_layers]
+            # evict order probably doesn't matter though?
+            evict_candidates = active_layers - set(fetch)
+
+            # evict_candidates isn't quite right-
+            # it thinks the currently active layer is evict-worthy since it doesn't show up in fetch.
+            # it's only saved by num_evict.  But i should figure out a cleaner way.
+            num_evict = max(
+                0, len(set(fetch + list(active_layers))) - max_active_layers
```
Comment (Skylion007) [inline torch/distributed/pipelining/schedules.py:862]: itertoools.chain here

File: torch/distributed/pipelining/schedules.py (hunk 1 above)
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:835]: yea i was being imprecise. i changed them all to 'stage'/'stages'

File: test/distributed/pipelining/test_schedule.py [hunk 4]
```diff
@@ -115,5 +122,54 @@ def test_pipeline_order(self, ScheduleClass):
...
+    @parametrize(
+        "test_info",
+        [
+            {
+                "compute": ["0F0", "0F1", "   ", "0B0", "0B1"],
+                "comms": ["0UNSHARD", "0F0", "0F1", "0B0", "0B1", "0RESHARD"],
+            },
+        ],
+    )
+    def test_unshard_reshard(self, test_info):
```
Comment (kwen2501) [inline test/distributed/pipelining/test_schedule.py:156]: nit: can you comment on purpose of test for better readability? Thanks!

File: test/distributed/pipelining/test_schedule.py [hunk 5]
```diff
@@ -10,8 +10,15 @@
...
 from torch.distributed.pipelining.schedules import (
+    _Action,
+    _add_unshard_reshard,
     _format_pipeline_order,
     _validate_pipeline_order,
+    B,
+    F,
+    RESHARD,
+    UNSHARD,
+    W,
```
Comment (kwen2501) [inline test/distributed/pipelining/test_schedule.py:21]: nit: can you make B, F, W more descriptive? Can be in a separate PR if it would involve larger changes. Thanks

File: test/distributed/pipelining/test_schedule.py [hunk 6]
```diff
@@ -115,5 +122,54 @@ def test_pipeline_order(self, ScheduleClass):
...
+        "action_str_and_ref",
+        [
+            ("1F0", _Action(1, F, 0)),
+            ("2B1", _Action(2, B, 1)),
+            ("0W3", _Action(0, W, 3)),
+            ("1UNSHARD", _Action(1, UNSHARD)),
+            ("3RESHARD", _Action(3, RESHARD)),
+        ],
+    )
+    def test_action_parse(self, action_str_and_ref):
```
Comment (kwen2501) [inline test/distributed/pipelining/test_schedule.py:141]: nit: can you comment on purpose of test for better readability? Thanks!

File: torch/distributed/pipelining/schedules.py [hunk 7]
```diff
@@ -53,15 +57,21 @@ def from_str(action):
...
 
 
 F = _ComputationType.FORWARD
 B = _ComputationType.BACKWARD
 W = _ComputationType.WEIGHT
+UNSHARD = _ComputationType.UNSHARD
+RESHARD = _ComputationType.RESHARD
 
-_action_regex = re.compile(r"(\d+)([F,B,W])(\d*)")
+_action_regex = re.compile(r"(\d+)([F,B,W]|UNSHARD|RESHARD)(\d*)")
```
Comment (kwen2501) [inline torch/distributed/pipelining/schedules.py:74]: nit: can you comment on what this regex does for better readability? Thanks!

File: torch/distributed/pipelining/schedules.py [hunk 8]
```diff
@@ -743,6 +753,99 @@ def _step_microbatches(
...
+    active_stages: Set[int] = set()
+    unshard_actions: List[Optional[_Action]] = []
+
+    def _unshard(stage_index: int):
+        active_stages.add(stage_index)
+        unshard_actions.append(_Action(stage_index, UNSHARD))
+
+    def _reshard(stage_index: int):
+        active_stages.remove(stage_index)
+        unshard_actions.append(_Action(stage_index, RESHARD))
```
Comment (kwen2501) [inline torch/distributed/pipelining/schedules.py:819]: Why using "unshard_actions" here? Just for saving one variable?

File: torch/distributed/pipelining/schedules.py [hunk 9]
```diff
@@ -743,6 +753,99 @@ def _step_microbatches(
...
+            evict,
+        )
+
+        for stage in evict:
+            _reshard(stage)
+        for stage in fetch:
+            _unshard(stage)
+        unshard_actions.append(action)
+
+    return unshard_actions
```
Comment (kwen2501) [inline torch/distributed/pipelining/schedules.py:846]: A general question: how does this interact with the code of FSDP? Which library would be the actual "firer" of the unshard ops?

**Review by kwen2501 — APPROVED** [review]
Code looks good to me. I only have a general question about interoperability with FSDP and some minor requests for comments.

File: torch/distributed/pipelining/schedules.py (hunk 9 above)
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:846]: I am working on a PR now that adds the 'runtime' component that would read this schedule format and execute the operations.  Probably what i'll do for the 'unshard' implementation is, 
 - if the stage module is an FSDP instance, call fsdp's unshard
 - otherwise, no-op.  (this way the unshard lowering can be run in all cases even if someone uses it with DDP..) 

If we need to decide at lowering time whether we're using FSDP vs DDP vs no DP, I think we can do that, but it makes composability a little more annoying so i'd rather not.

File: torch/distributed/pipelining/schedules.py (hunk 8 above)
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:819]: not sure what you mean about saving a variable.  unshard_actions is basically the 'new' version of the actions, which includes unshard/reshard.  the original actions list (compute only) is not modified.

File: test/distributed/pipelining/test_schedule.py (hunk 5 above)
Comment (wconstab) [reply test/distributed/pipelining/test_schedule.py:21]: Wdym, name them longer names?  I wanted them to be 1 character so they were easy to read/type in the instantiations below.  I'm happy to add more comments but hopefully you don't insist on writing it out as 'FORWARD' everywhere?

File: torch/distributed/pipelining/schedules.py [hunk 10]
```diff
@@ -36,12 +36,16 @@ class _ComputationType(Enum):
     FORWARD = 1
     BACKWARD = 2
     WEIGHT = 3
+    UNSHARD = 4
+    RESHARD = 5
```
Comment (H-Huang) [inline torch/distributed/pipelining/schedules.py:40]: High level question naming-wise since the next PR also adds `SEND_F`, `RECV_F`, etc. Should these and unshard/reshard be under a different enum like `_CommunicationType`?

These aren't really related to what the stage is computing and also might be easier to read code-wise to distinguish between the two. And maybe we can refactor the pipeline schedule order's naming to be something like `PipelineScheduleEvents` which encompass the computation and communication

File: torch/distributed/pipelining/schedules.py [hunk 11]
```diff
@@ -743,6 +753,76 @@ def _step_microbatches(
...
+        for a in next_actions:
+            if a is not None and a.stage_index not in seen:
+                seen.add(a.stage_index)
+                ret.append(a.stage_index)
+                if len(ret) == count:
+                    break
+        return ret
+
+    active_stages: Set[int] = set()
+    unshard_actions: List[Optional[_Action]] = []
```
Comment (H-Huang) [inline torch/distributed/pipelining/schedules.py:788]: This naming confused me at first, instead of `unshard_actions` i think this should be renamed like `fsdp_aware_compute_actions` or something

File: torch/distributed/pipelining/schedules.py [hunk 12]
```diff
@@ -743,6 +753,76 @@ def _step_microbatches(
...
+    """Given a basic schedule involving only compute actions (F,B,W), add UNSHARD/RESHARD actions for FSDP.
+
+    UNSHARD refers to fetching the full contents of an FSDP-sharded layer, requiring an all-gather operation.
+    RESHARD does the opposite, releasing memory (but doing no commmunication)
+
+    We abandon the "timestep lock"  during lowering
+
+    max_active_stages controls how many prefetches we allow. It should be measured in mb and tuneable but in practice
+    3 stages is probably the thing we want?
+    (to account for having one f and one b active, and something else prefetching?)
```
Comment (H-Huang) [inline torch/distributed/pipelining/schedules.py:769]: what did you mean by the "something else prefetching" comment?

**Review by H-Huang — APPROVED** [review]
Looks good to me

File: torch/distributed/pipelining/schedules.py (hunk 12 above)
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:769]: i'm not sure. i can't even remember why i thought we need 3 instead of 2 now. :/

File: torch/distributed/pipelining/schedules.py (hunk 10 above)
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:40]: > refactor the pipeline schedule order's naming to be something like PipelineScheduleEvents

this sgtm

> Should these and unshard/reshard be under a different enum like _CommunicationType

well in the next PR i initially separated them under _Action.computation_type and _Action.comm_type but I didn't like that and preferred to just have one 'action' field within action.  It should be renamed from _ComputationType to something else though.  Maybe _ActionType.

The reason i like them together instead of separate is becuase in the runtime we basically need a switch statement over the union of them, so having them come from 2 fields is a bit annoying.

I could probably also make _CommType and _CompType and have _Action.action: Union[_CommType, _CompType].  Wdyt?

Comment (H-Huang) [reply torch/distributed/pipelining/schedules.py:40]: > I could probably also make _CommType and _CompType and have _Action.action: Union[_CommType, _CompType]. Wdyt?

I like this. I also think this is more applicable later when we want to perform validation on the comms, then you past in the list of the comm types. Or even later if we do automatic bubble estimation, we pass in the list of the comp types. Then the full lowered schedule is a mix of both. I think that is intuitive!
